### PR TITLE
Apply license choices to SPDX expressions

### PR DESCRIPTION
**Pull request by MarcelBochtler**
This is a draft how license choices should be applied.
The signature of the tested function looks like this:
`fun applyChoice(choice: SpdxExpression, subtreeMatcher: SpdxExpression = this): SpdxExpression`

See #3396 and #2610

### CODE REVIEW

File: spdx-utils/src/test/kotlin/SpdxExpressionTest.kt [hunk 1]
```diff
@@ -367,6 +369,84 @@ class SpdxExpressionTest : WordSpec() {
...
+                }
+            }
+            
+            "return the expression if the matcher does not match the whole expression" {
+                val spdxExpression = "a OR b OR c".toSpdx()
+                val expressionMatcher = "b OR c".toSpdx()
+
+                val resolvedLicense = spdxExpression.applyChoice(expressionMatcher, "b".toSpdx())
+
+                resolvedLicense shouldBe "a OR b OR c".toSpdx()
```
Comment (MarcelBochtler) [inline spdx-utils/src/test/kotlin/SpdxExpressionTest.kt:419]: Not really sure about that one.
Is `b OR c` a valid subtree of `a OR b OR c` and therefore `b` would be a valid choice?

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 2]
```diff
@@ -153,6 +154,22 @@ sealed class SpdxExpression {
      */
     open fun offersChoice(): Boolean = false
 
+    /**
+     * Applies a license [choice] to the given SPDX [subtreeMatcher].
+     * The [subtreeMatcher] must not contain more than two choices.
+     */
+    open fun applyChoice(subtreeMatcher: SpdxExpression, choice: SpdxExpression): SpdxExpression {
```
Comment (mnonnenmacher) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:161]: I don't understand the API here, why is `choice` not applied to `this` but to `subtreeMatcher`? And why must it not contain more than two choices?

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 3]
```diff
@@ -491,6 +533,20 @@ data class SpdxLicenseReferenceExpression(
     override fun getLicenseUrl(): String? = null
 }
 
+/**
+ * Aggregates multiple SpdxExpressions to a [SpdxCompoundExpression] concatenated using [SpdxOperator.OR].
+ */
+fun List<SpdxExpression>.choicesToSpdx(): SpdxExpression {
```
Comment (mnonnenmacher) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:539]: This should go to `Extensions.kt`.

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 4]
```diff
@@ -491,6 +533,20 @@ data class SpdxLicenseReferenceExpression(
...
+    if (size == 1) return this[0]
+
+    var result: SpdxExpression = this[0]
+    this.subList(1, this.size).forEach {
+        result = result or it
+    }
+
+    return result
```
Comment (mnonnenmacher) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:547]: I think above code does not handle the case that the list is empty. This could be implemented as:
```
spdx.reduceOrNull { left, right -> left or right }
```

File: spdx-utils/src/test/kotlin/SpdxExpressionTest.kt [hunk 5]
```diff
@@ -367,6 +369,113 @@ class SpdxExpressionTest : WordSpec() {
...
+
+                assertThrows(InvalidLicenseChoiceException::class.java) {
+                    spdxExpression.applyChoice(spdxExpression, "b".toSpdx())
+                }
+            }
+
+            "throw an exception if the matcher does not match the expression and the choice is invalid" {
+                val spdxExpression = "a".toSpdx()
+
+                assertThrows(InvalidLicenseChoiceException::class.java) {
```
Comment (mnonnenmacher) [inline spdx-utils/src/test/kotlin/SpdxExpressionTest.kt:392]: Use `shouldThrow` from kotest instead.

File: spdx-utils/src/test/kotlin/SpdxExpressionTest.kt [hunk 6]
```diff
@@ -367,6 +369,113 @@ class SpdxExpressionTest : WordSpec() {
...
+                val spdxExpression = "(a OR b) AND c".toSpdx()
+                val matcher = "a OR b".toSpdx()
+
+                val resolvedLicense = spdxExpression.applyChoice(matcher, "a".toSpdx())
+
+                resolvedLicense shouldBe "a AND c".toSpdx()
+            }
+        }
+
+        "choicesToSpdx()" should {
```
Comment (mnonnenmacher) [inline spdx-utils/src/test/kotlin/SpdxExpressionTest.kt:459]: Add a test for the case that the list is empty.

File: spdx-utils/src/main/kotlin/SpdxExpression.kt (hunk 4 above)
Comment (mnonnenmacher) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:547]: Or even shorter: `spdx.reduceOrNull(SpdxExpression::or)`

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: spdx-utils/src/test/kotlin/SpdxExpressionTest.kt (hunk 1 above)
Comment (mnonnenmacher) [reply spdx-utils/src/test/kotlin/SpdxExpressionTest.kt]: Yes, I would say so.

File: spdx-utils/src/main/kotlin/SpdxExpression.kt (hunk 2 above)
Comment (MarcelBochtler) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:161]: > why is `choice` not applied to `this` but to `subtreeMatcher`?

Because the choice might only be a choice of a single subtree. E.g. (a OR b) AND c.
In my implementation the user would choose `b` for the subtree `a OR b` and the result would be `a AND c`.
Another approach would be to not use subtree matchers at all. E.g. if the choice is `b` it will be chosen no matter if the expression is `a OR b`, `a OR b OR c` or `(a OR b) AND c`.
I had the impression from our conversation and from https://github.com/oss-review-toolkit/ort/pull/2610 that this is the wanted behavior.

> And why must it not contain more than two choices?

Because I thought that `a OR b OR c` should be handled in two separate choices:
- `a OR b` choose `b`
- `b OR c` choose `b`
 Allowing three choices would end in even more combinations that need to be covered:
- `a OR b` choose `b`
- `b OR c` choose `b`
- `a OR b OR c` choose `c`

If you see a use case that this separation is required I'd agree to remove this limitation.

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 7]
```diff
@@ -236,6 +253,31 @@ class SpdxCompoundExpression(
             SpdxOperator.AND -> left.offersChoice() || right.offersChoice()
         }
 
+    override fun applyChoice(subtreeMatcher: SpdxExpression, choice: SpdxExpression): SpdxExpression {
+        if (!subtreeMatcher.isValidChoice(choice)) {
+            throw InvalidLicenseChoiceException(
+                "$choice is not a valid choice for the subtree expression. Valid choices are: ${validChoices()}."
+            )
+        }
+        if (subtreeMatcher.validChoices().size > 2) {
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:262]: Nit: Blank line before this one.

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 8]
```diff
@@ -236,6 +253,31 @@ class SpdxCompoundExpression(
...
+                "Subtree matcher $subtreeMatcher must not contain more than two choices."
+            )
+        }
+
+        return replaceMatcherWithChoice(subtreeMatcher, choice)
+    }
+
+    private fun replaceMatcherWithChoice(subtreeMatcher: SpdxExpression, choice: SpdxExpression) =
+        if (toString().contains(subtreeMatcher.toString())) {
+            toString().replace(subtreeMatcher.toString(), choice.toString()).toSpdx()
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:273]: Maybe avoid calling `toString()` on the same objects multiple times by introducing variables?

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 9]
```diff
@@ -491,6 +533,20 @@ data class SpdxLicenseReferenceExpression(
     override fun getLicenseUrl(): String? = null
 }
 
+/**
+ * Aggregates multiple SpdxExpressions to a [SpdxCompoundExpression] concatenated using [SpdxOperator.OR].
+ */
+fun List<SpdxExpression>.choicesToSpdx(): SpdxExpression {
+    if (size == 1) return this[0]
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:540]: `first()` instead of `this[0]`? Same below.

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 10]
```diff
@@ -491,6 +533,20 @@ data class SpdxLicenseReferenceExpression(
...
+
+    var result: SpdxExpression = this[0]
+    this.subList(1, this.size).forEach {
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:543]: We prefer to drop `this.` if not strictly required.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: spdx-utils/src/main/kotlin/SpdxExpression.kt (hunk 9 above)
Comment (MarcelBochtler) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:540]: Not relevant anymore

File: spdx-utils/src/main/kotlin/SpdxExpression.kt (hunk 10 above)
Comment (MarcelBochtler) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:543]: Not relevant anymore

File: spdx-utils/src/main/kotlin/Extensions.kt [hunk 11]
```diff
@@ -101,3 +101,12 @@ fun SpdxLicense.toExpression(): SpdxLicenseIdExpression {
  */
 fun String.toSpdx(strictness: Strictness = Strictness.ALLOW_ANY): SpdxExpression =
     SpdxExpression.parse(this, strictness)
+
+/**
+ * Aggregates multiple SpdxExpressions to a [SpdxCompoundExpression] concatenated using [SpdxOperator.OR].
+ */
+fun List<SpdxExpression>.choicesToSpdx(): SpdxExpression {
+    require(this.isNotEmpty())
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/Extensions.kt:109]: Please drop `this.` here and below.

File: spdx-utils/src/main/kotlin/Extensions.kt [hunk 12]
```diff
@@ -101,3 +101,12 @@ fun SpdxLicense.toExpression(): SpdxLicenseIdExpression {
...
+
+    return this.reduceOrNull(SpdxExpression::or)!!
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/Extensions.kt:111]: As you already assert the list to be non-empty above, we can avoid the ugly non-null assertion here by using `reduce(SpdxExpression::or)`.

Comment (sschuberth) [reply spdx-utils/src/main/kotlin/Extensions.kt:111]: Or we could even make it a one-liner like

    fun List<SpdxExpression>.choicesToSpdx(): SpdxExpression =
        requireNotNull(reduceOrNull(SpdxExpression::or))

File: spdx-utils/src/main/kotlin/Extensions.kt [hunk 13]
```diff
@@ -101,3 +101,12 @@ fun SpdxLicense.toExpression(): SpdxLicenseIdExpression {
  */
 fun String.toSpdx(strictness: Strictness = Strictness.ALLOW_ANY): SpdxExpression =
     SpdxExpression.parse(this, strictness)
+
+/**
+ * Aggregates multiple SpdxExpressions to a [SpdxCompoundExpression] concatenated using [SpdxOperator.OR].
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/Extensions.kt:106]: This should use the imperative and say "Aggregate" (I just realized it's also wrong in the existing function docs above).

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 14]
```diff
@@ -153,6 +154,22 @@ sealed class SpdxExpression {
      */
     open fun offersChoice(): Boolean = false
 
+    /**
+     * Applies a license [choice] to the given SPDX [subtreeMatcher].
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:158]: Imperative "Apply".

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 15]
```diff
@@ -236,6 +253,37 @@ class SpdxCompoundExpression(
...
+    private fun replaceMatcherWithChoice(subtreeMatcher: SpdxExpression, choice: SpdxExpression): SpdxExpression {
+        val expressionString = toString()
+        val matcherString = subtreeMatcher.toString()
+        val choiceString = choice.toString()
+
+        return if (expressionString.contains(matcherString)) {
+            expressionString.replace(matcherString, choiceString).toSpdx()
+        } else {
+            val dismissedLicense = subtreeMatcher.validChoices().first { it != choice }
+            val remains = this.validChoices().filter { it != dismissedLicense }
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:281]: Drop `this.`. Can you maybe think of a better variable name than `remains`?

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: spdx-utils/src/main/kotlin/SpdxExpression.kt (hunk 15 above)
Comment (MarcelBochtler) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:281]: I tried ^^

Comment (MarcelBochtler) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt]: Removed the this. Also found an IntelliJ inspection to warn about redundant `this`.

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 16]
```diff
@@ -236,6 +253,37 @@ class SpdxCompoundExpression(
...
+    private fun replaceMatcherWithChoice(subtreeMatcher: SpdxExpression, choice: SpdxExpression): SpdxExpression {
+        val expressionString = toString()
+        val matcherString = subtreeMatcher.toString()
+        val choiceString = choice.toString()
+
+        return if (expressionString.contains(matcherString)) {
+            expressionString.replace(matcherString, choiceString).toSpdx()
+        } else {
+            val dismissedLicense = subtreeMatcher.validChoices().first { it != choice }
+            val notChosenLicenses = validChoices().filter { it != dismissedLicense }
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:281]: Nit: `unchosenLicenses`?

**Review by sschuberth — COMMENTED** [review]
I'm basically fine with the formal aspects of the PR now, but I'd like @mnonnenmacher to approve on the API.

File: spdx-utils/src/main/kotlin/SpdxExpression.kt (hunk 2 above)
Comment (sschuberth) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:161]: TBH, I had a different understanding of how choices should work. In my view, a valid choice should always be checked against the *whole* expression. There should be no such thing as choices on sub-expressions.

That is, if the original expression would be `(a OR b) AND c`, then `b` would not be a valid choice, as it omits `c`. Valid choices would be `a AND c` or `b AND c`.

Comment (MarcelBochtler) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:161]: @mnonnenmacher do you agree with @sschuberth?
I think this would make the API and the implementation easier and the configuration for users more straight forward.

Comment (fviernau) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:161]: I'm a bit surprise that making choices for sub-trees is being questioned again, as we discussed this in the meeting and I believe it became clear that it should be supported.

Picking up Sebastian's example `(a OR b) AND c` I believe the choice `(a OR b) ->  b` should be applicable to it and resolve it to `b AND c´.

Comment (sschuberth) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:161]: I guess I'm simply confused about the API, even after looking at the tests, mostly due to variable naming I believe. I *can* see a value of being able to first "narrow down" the choice, and incrementally doing that to get a final choice which doe snot contain an `OR` anymore.

What would help at least me for a better understanding would be to change the function as follows:

    /**
     * Apply a license [choice], optionally limited to the given [subExpression], and return an [SpdxExpression] where
     * the choice is resolved.
     */
    open fun applyChoice(choice: SpdxExpression, subExpression: SpdxExpression = this): SpdxExpression {

Comment (sschuberth) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:161]: > > And why must it not contain more than two choices?
> 
> Because I thought that `a OR b OR c` should be handled in two separate choices:

IMO, ideally the API would work in a way that I can generically say "whenever I come across sub-expression *X*, choose *Y*", so I should be able to say "whenever I come across sub-expression `a OR b OR c`, choose `b`" or also "whenever I come across sub-expression `a OR b OR c`, choose `b OR c`".

Comment (fviernau) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:161]: > IMO, ideally the API would work in a way that I can generically say "whenever I come across sub-expression _X_, choose _Y_", so I should be able to say "whenever I come across sub-expression `a OR b OR c`, choose `b`" or also "whenever I come across sub-expression `a OR b OR c`, choose `b OR c`".

Fully agree, that seem 100% inline with what I envisioned as well.

Comment (MarcelBochtler) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt:161]: I changed the API to your suggestions and tried to improve the readability.
Please also have a look at the tests that cover everything that I thought of. If a test doesn't exist I forgot that specific use case.

Short summary how the API will behave:
- Whenever I come across sub-expression `a OR b OR c`, choose `b`
- Whenever I come across sub-expression `a OR b OR c`, choose `b OR c`
- If the sub-expression does not match the expression, the expression is returned (should we use 
- Throws an exception if the choice is not valid for the subExpression

Comment (MarcelBochtler) [reply spdx-utils/src/main/kotlin/SpdxExpression.kt]: @fviernau, @sschuberth please have another look at the changes

File: spdx-utils/src/test/kotlin/SpdxExpressionTest.kt [hunk 17]
```diff
@@ -367,6 +368,118 @@ class SpdxExpressionTest : WordSpec() {
...
+            }
+
+            "throw an exception if the chosen license is not an valid option" {
+                val expression = "a OR b".toSpdx()
+                val choice = "c".toSpdx()
+
+                shouldThrow<InvalidLicenseChoiceException> { expression.applyChoice(choice) }
+            }
+
+            "apply the choice if the expression is not in dnf" {
```
Comment (sschuberth) [inline spdx-utils/src/test/kotlin/SpdxExpressionTest.kt:433]: Only a nit: Please use "DNF" here and in the two places below.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]
@fviernau @mnonnenmacher I believe this is the API we agreed on and now finally ready. Please also do a re-review so we're sure you're fine with the API and we can take it into use without expecting it to change again soon. Thanks in advance!

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 18]
```diff
@@ -236,6 +249,35 @@ class SpdxCompoundExpression(
             SpdxOperator.AND -> left.offersChoice() || right.offersChoice()
         }
 
+    override fun applyChoice(choice: SpdxExpression, subExpression: SpdxExpression): SpdxExpression {
+        if (!subExpression.validChoices().containsAll(choice.validChoices())) {
+            throw InvalidLicenseChoiceException(
+                "$choice is not a valid choice for the subtree expression. Valid choices are: ${validChoices()}."
```
Comment (mnonnenmacher) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:255]: "subexpression" instead of "subtree expression", and I think it would be good to also print the subexpression in this message. Or maybe just say "...is not a valid choice for $subExpression. ..."

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 19]
```diff
@@ -236,6 +249,35 @@ class SpdxCompoundExpression(
...
+            )
+        }
+
+        return replaceMatcherWithChoice(subExpression, choice)
+    }
+
+    private fun replaceMatcherWithChoice(subExpression: SpdxExpression, choice: SpdxExpression): SpdxExpression {
```
Comment (mnonnenmacher) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:262]: Is "matcher" a left-over from a renaming? Maybe rename the function to `replaceSubexpressionWithChoice`.

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 20]
```diff
@@ -236,6 +249,35 @@ class SpdxCompoundExpression(
...
+        val expressionString = toString()
+        val matcherString = subExpression.toString()
```
Comment (mnonnenmacher) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:264]: As above, rename to `subexpressionString`?

File: spdx-utils/src/main/kotlin/SpdxExpression.kt [hunk 21]
```diff
@@ -516,3 +558,5 @@ enum class SpdxOperator(
      */
     OR(0)
 }
+
+class InvalidLicenseChoiceException(message: String) : Exception(message)
```
Comment (mnonnenmacher) [inline spdx-utils/src/main/kotlin/SpdxExpression.kt:562]: Would it make sense to extend `SpdxException` instead?

File: spdx-utils/src/test/kotlin/SpdxExpressionTest.kt [hunk 22]
```diff
@@ -367,6 +368,118 @@ class SpdxExpressionTest : WordSpec() {
...
+                val choice = "b".toSpdx()
+
+                shouldThrow<InvalidLicenseChoiceException> { expression.applyChoice(choice) }
+            }
+
+            "return the choice if the subExpression contains multiple choices" {
+                val expression = "a OR b OR c".toSpdx()
+                val choice = "a".toSpdx()
+
+                val result = expression.applyChoice(choice)
```
Comment (mnonnenmacher) [inline spdx-utils/src/test/kotlin/SpdxExpressionTest.kt:392]: Currently this is the same test as "apply the choice if the expression contains multiple choices" below, according to the test name I think you wanted to pass a subexpression here.

File: spdx-utils/src/test/kotlin/SpdxExpressionTest.kt [hunk 23]
```diff
@@ -367,6 +368,118 @@ class SpdxExpressionTest : WordSpec() {
...
+            "apply the choice if the expression contains multiple choices" {
+                val expression = "a OR b OR c".toSpdx()
+                val choice = "b".toSpdx()
+
+                val result = expression.applyChoice(choice)
+
+                result shouldBe "b".toSpdx()
+            }
+
+            "throw an exception if the chosen license is not an valid option" {
```
Comment (mnonnenmacher) [inline spdx-utils/src/test/kotlin/SpdxExpressionTest.kt:426]: Typo: "a valid"

File: spdx-utils/src/test/kotlin/SpdxExpressionTest.kt [hunk 24]
```diff
@@ -367,6 +368,118 @@ class SpdxExpressionTest : WordSpec() {
...
+
+                result shouldBe "a AND c AND d OR b AND c AND d OR b AND c AND e".toSpdx()
+            }
+
+            "return the expression if the subExpression does not match" {
+                val expression = "a OR b OR c".toSpdx()
+                val choice = "x".toSpdx()
+                val subExpression = "x OR y OR z".toSpdx()
+
+                val result = expression.applyChoice(choice, subExpression)
```
Comment (mnonnenmacher) [inline spdx-utils/src/test/kotlin/SpdxExpressionTest.kt:457]: Shouldn't it throw an exception if the subexpression is invalid like for invalid license choices?

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

**Comment by mnonnenmacher** [comment]
> @fviernau @mnonnenmacher I believe this is the API we agreed on and now finally ready. Please also do a re-review so we're sure you're fine with the API and we can take it into use without expecting it to change again soon. Thanks in advance!

I'm mostly fine with the API and had only minor comments. The only API change I mentioned in my comments is if we should throw an exception if `applyChoice` is called with an invalid `subExpression` like we do for invalid choices.

File: spdx-utils/src/test/kotlin/SpdxExpressionTest.kt (hunk 22 above)
Comment (MarcelBochtler) [reply spdx-utils/src/test/kotlin/SpdxExpressionTest.kt:392]: It was a leftover from before the API change where not more than one choice were allowed. As now choices can be applied I removed this test.

File: spdx-utils/src/test/kotlin/SpdxExpressionTest.kt (hunk 24 above)
Comment (MarcelBochtler) [reply spdx-utils/src/test/kotlin/SpdxExpressionTest.kt:457]: Yes I agree that this would make the API more consistent and will fail early if the developer uses this API in another way as it was intended.

**Review by sschuberth — APPROVED** [review]
Also need @mnonnenmacher to re-review and approve.

**Review by mnonnenmacher — APPROVED** [review]

**Comment by sschuberth** [comment]
Merging this despite the stuck *markdown-link-check* as all other checks passed.
